### PR TITLE
Hubspot Marketing

### PR DESCRIPTION
**Pull request by troystaylor**
---
When submitting a connector, please check the following conditions for your PR.

- [X] I attest that the connector works and I verified by deploying and testing all the operations.
- [X] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [X] I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`.

### CODE REVIEW

**Comment by mamurshe** [comment]
[[certify-connector]]

**Comment by mamurshe** [comment]
[[certify-connector]]

**Comment by mamurshe** [comment]
[[certify-connector]]

**Comment by mamurshe** [comment]
[[certify-connector]]

**Review by mamurshe — CHANGES_REQUESTED** [review]
Your connection parameter definition was completely broken. I fixed some of it and needs more fix. I am including a sample. Can you please fix both HubSpot Marketing and HubSpot CMS?
```
    "connectionParameters": {
      "token": {
        "type": "oauthSetting",
        "oAuthSettings": {
          "identityProvider": "[variables('identityProvider')]",
          "clientId": "[variables('clientIdAlt')]",
          "clientSecret": "[parameters('clientSecret')]",
          "scopes": [],
          "redirectMode": "[variables('PerRegionSettings')[parameters('apimAccountName')].redirectMode]",
          "customParameters": {
            "TenantId": {},
            "ResourceUriAAD": {
              "value": "[variables('resourceUriAAD')]"
            },
            "resourceUri": {
              "value": "[variables('resourceURIAlt')]"
            },
            "loginUri": {
              "value": "[variables('loginUriAAD')]"
            },
            "loginUriAAD": {
              "value": "[variables('loginUriAAD')]"
            }
          }
        }
      },
      "token:TenantId": {
        "type": "string",
        "metadata": {
          "sourceType": "AzureActiveDirectoryTenant"
        },
        "uiDefinition": {
          "displayName": "Tenant",
          "description": "The tenant ID of for the Azure Active Directory application",
          "constraints": {
            "required": "false",
            "hidden": "true"
          }
        }
      }
    },
    "connectionParameterSets": {
      "uiDefinition": {
        "displayName": "Authentication Type",
        "description": "Type of authentication to be used."
      },
      "values": [
        {
          "name": "firstParty",
          "uiDefinition": {
            "displayName": "First Party",
            "description": "Use First Party Authentication.",
            "constraints": {
              "hidden": "false"
            }
          },
          "metadata": {
            "allowSharing": false
          },
          "parameters": {
            "token": {
              "type": "oauthSetting",
              "oAuthSettings": {
                "identityProvider": "aadcertificate",
                "clientId": "[variables('clientId')]",
                "scopes": [],
                "redirectMode": "[variables('redirectMode')]",
                "customParameters": {
                  "TenantId": {},
                  "ResourceUriAAD": {
                    "value": "[variables('resourceURI')]"
                  },
                  "resourceUri": {
                    "value": "[variables('resourceURI')]"
                  },
                  "loginUri": {
                    "value": "[variables('loginUriAAD')]"
                  },
                  "loginUriAAD": {
                    "value": "[variables('loginUriAAD')]"
                  }
                }
              },
              "uiDefinition": {
                "displayName": "Token",
                "description": "Token for authentication.",
                "constraints": {
                  "required": "true",
                  "hidden": "false"
                }
              }
            },
            "token:TenantId": {
              "type": "string",
              "metadata": {
                "sourceType": "AzureActiveDirectoryTenant"
              },
              "uiDefinition": {
                "displayName": "Tenant",
                "description": "The tenant ID of for the Azure Active Directory application",
                "constraints": {
                  "required": "false",
                  "hidden": "true"
                }
              }
            }
          }
        },
        {
          "name": "thirdParty",
          "uiDefinition": {
            "displayName": "Third Party (Deprecated)",
            "description": "Use Third Party Authentication (Deprecated).",
            "constraints": {
              "hidden": "true"
            }
          },
          "metadata": {
            "allowSharing": false
          },
          "parameters": {
            "token": {
              "type": "oauthSetting",
              "oAuthSettings": {
                "identityProvider": "[variables('identityProvider')]",
                "clientId": "[variables('clientIdAlt')]",
                "clientSecret": "[parameters('clientSecret')]",
                "scopes": [],
                "redirectMode": "[variables('PerRegionSettings')[parameters('apimAccountName')].redirectMode]",
                "customParameters": {
                  "TenantId": {},
                  "ResourceUriAAD": {
                    "value": "[variables('resourceUriAAD')]"
                  },
                  "resourceUri": {
                    "value": "[variables('resourceURIAlt')]"
                  },
                  "loginUri": {
                    "value": "[variables('loginUriAAD')]"
                  },
                  "loginUriAAD": {
                    "value": "[variables('loginUriAAD')]"
                  }
                }
              },
              "uiDefinition": {
                "displayName": "Token",
                "description": "Token for authentication.",
                "constraints": {
                  "required": "true",
                  "hidden": "false"
                }
              }
            },
            "token:TenantId": {
              "type": "string",
              "metadata": {
                "sourceType": "AzureActiveDirectoryTenant"
              },
              "uiDefinition": {
                "displayName": "Tenant",
                "description": "The tenant ID of for the Azure Active Directory application",
                "constraints": {
                  "required": "false",
                  "hidden": "true"
                }
              }
            }
          }
        }
      ]
    },

```

**Comment by troystaylor** [comment]
@mamurshe Can you explain why I need to use variables and parameters like the example you gave? I don't use them elsewhere in the connector

**Comment by troystaylor** [comment]
Updated apiProperties

**Review by mamurshe — CHANGES_REQUESTED** [review]
When I try to create a custom connector using these asset with paconn CLI I get an error. Can you please update the apiProperties.json using `paconn download` command?

```
independent-publisher-connectors\HubSpot Marketing>paconn create -d apiDefinition.swagger.json -p apiProperties.json -e 5827fea3-d788-43a6-9861-3c82de256bce
'properties'
Traceback (most recent call last):
  File "c:\python\lib\site-packages\knack\cli.py", line 206, in invoke
    cmd_result = self.invocation.execute(args)
  File "c:\python\lib\site-packages\knack\invocation.py", line 188, in execute
    cmd_result = parsed_args.func(params)
  File "c:\python\lib\site-packages\knack\commands.py", line 131, in __call__
    return self.handler(*args, **kwargs)
  File "c:\python\lib\site-packages\knack\commands.py", line 238, in _command_handler
    result = op(client, **command_args) if client else op(**command_args)
  File "c:\python\lib\site-packages\paconn\commands\create.py", line 47, in create
    connector_id = upsert(
  File "c:\python\lib\site-packages\paconn\operations\upsert.py", line 82, in upsert
    properties = property_definition[_PROPERTIES]
KeyError: 'properties'
```

**Review by mamurshe — CHANGES_REQUESTED** [review]
"redirectUrl": ["https://global.consent.azure-apim.net/redirect", "https://global-test.consent.azure-apim.net/redirect"],
to
"redirectUrl": "https://global.consent.azure-apim.net/redirect",

**Review by mamurshe — CHANGES_REQUESTED** [review]
Can you please add `connectionParameters` section as shown below? It's needed for deployment.

```
  "properties": {
    "connectionParameters": {
      "api_key": {
        "type": "securestring",
        "uiDefinition": {
          "constraints": {
            "clearText": false,
            "required": "true",
            "tabIndex": 3
          },
          "schema": {
            "description": "Enter your API Key for HubSpot",
            "type": "securestring"
          },
          "displayName": "API Key generated in HubSpot"
        }
      }
    },
    "connectionParameterSets": {
```

File: independent-publisher-connectors/HubSpot Marketing/apiProperties.json [hunk 1]
```diff
@@ -0,0 +1,75 @@
+{
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot Marketing/apiProperties.json:1]: Please remove apikey auth
```
{
  "properties": {
    "connectionParameters": {
      "token": {
        "type": "oauthSetting",
        "oAuthSettings": {
          "identityProvider": "oauth2",
          "clientId": "[Your_Value]",
          "scopes": [
            "forms files hubdb transactional-email settings.users.write conversations.visitor_identification.tokens.create integration-sync tickets e-commerce files.ui_hidden.read crm.objects.marketing_events.read accounting crm.objects.marketing_events.write crm.schemas.custom.read crm.objects.custom.read crm.objects.feedback_submissions.read sales-email-read crm.objects.custom.write forms-uploaded-files settings.users.read crm.import contacts content reports social automation settings.billing.write actions timeline business-intelligence oauth"
          ],
          "redirectMode": "Global",
          "redirectUrl": "https://global.consent.azure-apim.net/redirect",
          "properties": {
            "IsFirstParty": "False"
          },
          "customParameters": {
            "authorizationUrl": {
              "value": "https://app.hubspot.com/oauth/authorize"
            },
            "tokenUrl": {
              "value": "https://api.hubapi.com/oauth/v1/token"
            },
            "refreshUrl": {
              "value": "https://api.hubapi.com/oauth/v1/token"
            }
          }
        }
      }
    },
    "iconBrandColor": "#da3b01",
    "capabilities": [],
    "publisher": "Hitachi Solutions"
  }
}
```

File: independent-publisher-connectors/HubSpot Marketing/apiDefinition.swagger.json [hunk 2]
```diff
@@ -0,0 +1,1455 @@
...
+            }
+          }
+        ]
+      }
+    }
+  },
+  "definitions": {},
+  "parameters": {},
+  "responses": {},
+  "securityDefinitions": {
```
Comment (mamurshe) [inline independent-publisher-connectors/HubSpot Marketing/apiDefinition.swagger.json:1422]: Please combine `securityDefinitions` and `security` sections.

**Review by mamurshe — CHANGES_REQUESTED** [review]

**Pull request merged**